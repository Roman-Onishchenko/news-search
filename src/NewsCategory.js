import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as actions from './actions';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import './styles/NewsList.css';
import preload from './images/preview.png';

const Header = () => (
  <header>
    <nav>
      <ul className="header-nav">
        <li><Link to='/'>Назад</Link></li>
      </ul>
    </nav>
  </header>
);

export class NewsCategory extends Component {

  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const source = this.props.match.params.source;
    this.props.fetchNews(source);
    this.props.preloadtoTrue();
  }

  render() {

    const newsList = this.props.newslist.get('foundNews').toJS();
    const preview = this.props.newslist.get('preload');

    return (
      <div>

        {preview && <div className="preloader">
          <img className="preload_img" src={preload} alt="preload" />
        </div>}

        <div className="news-category">

          <Header />
          <ul>
            {newsList.map((val, index) =>
              <li key={index}
                  className="news-item"
              >
                <p className="news-item__title">{val.title}</p>
                <img className="news-item__img"
                  src={val.urlToImage}
                  alt="news" />
                <span className="news-item__author">{val.author} at </span>
                <span className="news-item__date">{new Date(val.publishedAt).toDateString()}</span>
                <p className="news-item__description">{val.description}</p>
                <a className="news-item__link"
                  target="_blank"
                  href={val.url}>Visit official website</a>
              </li>
              )
            }
          </ul>
        </div>
      </div>
    )
  };
};

NewsCategory.propTypes = {
  newslist: PropTypes.object
};

export default connect(
  state => ({
    newslist: state.newsData,
  }),
  dispatch => bindActionCreators(actions, dispatch)
)(NewsCategory);
