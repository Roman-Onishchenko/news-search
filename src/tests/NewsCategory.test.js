import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount, render } from 'enzyme';
import sinon from 'sinon';
import { Route, Link, MemoryRouter } from 'react-router-dom';
import { NewsCategory } from '../NewsCategory';
import { Map, List } from 'immutable';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

describe('NewsCategory', () => {

  it('renders without crashing', () => {

    const fetchNews = sinon.spy();
    const preloadtoTrue = sinon.spy();

    const newslist = new Map({
       foundNews: new List(),
       preload: true
    });

    const match = {
      params: {
        source: "path"
      }
    };

    const renderedComponent = shallow(
        <NewsCategory
          fetchNews={fetchNews}
          newslist={newslist}
          preloadtoTrue={preloadtoTrue}
          match = {match}
        />
    );

    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('launches news fetch', () => {
    const fetchNews = sinon.spy();
    const preloadtoTrue = sinon.spy();

    const newslist = new Map({
       foundNews: new List(),
       preload: true
    });

    const preview = new Map({
       preload: true
    });

    const match = {
      params: {
        source: "path"
      }
    };

    const wrapper = mount(
      <MemoryRouter>
        <NewsCategory
          fetchNews={fetchNews}
          newslist={newslist}
          preloadtoTrue={preloadtoTrue}
          match = {match}
        />
    </MemoryRouter>);

    expect(fetchNews.calledOnce).toBeTruthy();
    expect(preloadtoTrue.calledOnce).toBeTruthy();
  });

  it('checks items rendering', () => {

    const fetchNews = sinon.spy();
    const preloadtoTrue = sinon.spy();

    const newslist = new Map({
       foundNews:[{
         id: 1,
         title: 'News'
       },

       {
         id: 2,
         title: 'News'
       },

       {
         id: 3,
         title: 'News'
       }
     ],
       preload: true
    });

    const preview = new Map({
       preload: true
    });

    const match = {
      params: {
        source: "path"
      }
    };

    const wrapper = shallow(

        <NewsCategory
          preview={preview}
          fetchNews={fetchNews}
          newslist={newslist}
          preloadtoTrue={preloadtoTrue}
          match={match}
        />

    );

    const listArr = wrapper.find('li');
    expect(listArr).toHaveLength(newslist.get('foundNews').length);
  });

});
