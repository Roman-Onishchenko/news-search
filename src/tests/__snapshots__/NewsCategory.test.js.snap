// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsCategory renders without crashing 1`] = `
<div>
  <div
    className="preloader"
  >
    <img
      alt="preload"
      className="preload_img"
      src="preview.png"
    />
  </div>
  <div
    className="news-category"
  >
    <Header />
    <ul />
  </div>
</div>
`;
